draft-ietf-dnsop-svcb-https.md

```diff
@@ -495,59 +495,53 @@ is the effective TargetName:
-   plus appropriate prefixes for the scheme (see {{svcb-names}}).
+1. Let $QNAME be the service name plus appropriate prefixes for the
+   scheme (see {{svcb-names}}).
 
-2. In parallel, issue AAAA/A queries for $ADDR_QNAME
-   and a SVCB query for $SVCB_QNAME.
-   The answers for these may or may not include CNAME pointers
-   before reaching one or more of these records.
+2. Issue a SVCB query for $QNAME.
 
-3. If an AliasMode SVCB record is returned for $SVCB_QNAME,
-   clients MUST set $ADDR_QNAME and $SVCB_QNAME to its TargetName (without
+3. If an AliasMode SVCB record is returned, set $QNAME to its TargetName (without
    additional prefixes) and loop back to step 2,
    subject to chain length limits and loop detection heuristics (see
    {{client-failures}}).
 
-4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned
-   for $SVCB_QNAME, clients SHOULD select the highest-priority compatible record.
-   This record's TargetName and SvcParams represent the preferred endpoint.  If
-   connection to this endpoint fails, the client SHOULD try to connect using
-   values from the next-highest-priority compatible record, etc. If all attempts
-   fail, clients SHOULD go to step 5 (except as noted in {{ech-client-behavior}}).
+4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned,
+   these represent the alternative endpoints.
 
-5. At this point there are no usable ServiceMode records, because
-   - there were no SVCB records found for $SVCB_QNAME, OR
-   - all records found were incompatible with this client, OR
-   - all connection attempts using ServiceMode records failed.
-
-   Accordingly, clients SHALL connect to the
-   endpoint consisting of $ADDR_QNAME, the authority endpoint's
-   port number, and no SvcParams.
-
-If all of the above connection attempts fail, clients MAY connect to the
-authority endpoint (except as noted in {{client-failures}} and
-{{ech-client-behavior}}).
+5. Otherwise, SVCB resolution has failed.
 
 This procedure does not rely on any recursive or authoritative DNS server to
 comply with this specification or have any awareness of SVCB.
 
-When selecting between AAAA and A records to use, clients may use an approach
-such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.
+Once SVCB resolution is completed, the client proceeds with connection
+establishment.  Clients SHOULD try higher-priority alternatives first, with
+fallback to lower-priority alternatives.  Clients MUST issue AAAA and/or A
+queries for the selected TargetName, and MAY choose between them using an
+approach such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.
+
+A client is called "SVCB-reliant" if it connects only to the indicated
+ServiceMode endpoints, and "SVCB-optional" otherwise.  Clients for pre-existing
+protocols SHALL implement SVCB-optional behavior (except as noted in
+{{client-failures}} and {{ech-client-behavior}}).
```

Comment:
This definition of SVCB-reliant doesn't match with my understanding.  I think that what you want to do is note that there are two reasons to make an A/AAAA query in parallel.  The primary one being that the client might be willing to fall back to A/AAAA queries if there are no SVCB records, but you might also want to allow for the possibility that a TargetName form SVCB record references the same name and so there might be an advantage in making the A/AAAA queries in parallel to allow for that outcome.

(Unfortunately, if you *need* SVCB, you need the TargetName form, and you don't get to predict when that will appear, so you basically have to make the parallel query always once you decide that it is worth the parallel.)

You mention the latter below, which is good, but I think that it applies here.

Reply:
I'm not sure I understand your concern here, but I hope these changes have clarified it somewhat.

Reply:
I think that the key difference between reliant and optional is whether the information in the ServiceMode is critical to operation of the client.

Related question: does an SVCB-optional client switch to SVCB-reliant if it encounters an AliasMode record?  Or, can it follow to the target name and just consume A/AAAA (or whatever) instead, without ever having seen a ServiceMode record?

Reply:
> I think that the key difference between reliant and optional is whether the information in the ServiceMode is critical to operation of the client.

Agreed.  How does that differ from the definition given here?

> Related question: does an SVCB-optional client switch to SVCB-reliant if it encounters an AliasMode record?

No.

> Or, can it follow to the target name and just consume A/AAAA (or whatever) instead, without ever having seen a ServiceMode record?

Yes.  That's important, to support use of AliasMode pointing to a hosting provider that doesn't have a SVCB record.

Additionally, as currently specified, SVCB-optional clients are generally encouraged (and in some cases required) to fall back to non-SVCB connection if all the SVCB alternative endpoints are tried and they all fail. (See #256 for some debate about that.)

Reply:
>  How does that differ from the definition given here?

The definition here concentrates on what the client does (connect to the ServiceForm endpoint), rather than the reasons for the same.

Does the draft explain the AliasForm-CNAME isomorphism?  There's a lot of text, but I can't see anything there.  Seems important.

Reply:
> The definition here concentrates on what the client does (connect to the ServiceForm endpoint), rather than the reasons for the same.

OK, I've reversed the polarity of the definitions.  Does that help?

> Does the draft explain the AliasForm-CNAME isomorphism?

It's not really an isomorphism.  [The AliasMode section](https://tools.ietf.org/html/draft-ietf-dnsop-svcb-https-02#section-2.4.2) discusses the relationship in some detail.  I'm happy to improve the description but I think it's out of scope for this PR.